### PR TITLE
Generating simulation files from a template

### PR DESCRIPTION
**Pull request by whytheplatypus**
This allows the simulator to use templates to generate files of any size. It also includes two template functions for generating a random string or a random int.

### CODE REVIEW

File: tests/bad-uploader/template.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	slog.Debug("reading template")
+	n, err := tg.r.Read(p)
+	slog.Debug("read template")
+	if err == io.EOF {
+		if tg.Repeats < 1 {
+			return 0, err
+		}
+		tg.Repeats--
+		slog.Debug("hit eof")
+		//TODO if we should stop return the EOF?
```
Comment (whytheplatypus) [inline tests/bad-uploader/template.go:90]:
```suggestion
```

File: tests/bad-uploader/template.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		if err := tg.next(); err != nil {
+			return n, err
+		}
+		// do we need to re-read here or will it just try again?
```
Comment (whytheplatypus) [inline tests/bad-uploader/template.go:94]:
```suggestion
```

File: tests/bad-uploader/main.go [hunk 3]
```diff
@@ -216,7 +231,20 @@ func main() {
 	}
 
 	tStart := time.Now()
-	if load > 0 {
+	if duration > 0 {
+		slog.Info("Running duration test test", "duration", duration)
```
Comment (cfarmer-fearless) [inline tests/bad-uploader/main.go:235]: test test

File: tests/bad-uploader/main.go [hunk 4]
```diff
@@ -264,20 +283,39 @@ func asPallelBenchmark(c chan TestCase, next func() TestCase) func(*testing.B) {
...
 	}
 
 	// create the tus client.
 	tusConf := tus.DefaultConfig()
 	tusConf.ChunkSize = int64(t.Chunk)
+	tusConf.HttpClient = &http.Client{}
 	if conf.tokenSource != nil {
 		tusConf.HttpClient = oauth2.NewClient(context.TODO(), conf.tokenSource)
 	}
+	tusConf.Header.Set("Upload-Defer-Length", "1")
```
Comment (cfarmer-fearless) [inline tests/bad-uploader/main.go:317]: What are these headers for?

File: tests/bad-uploader/main.go [hunk 5]
```diff
@@ -264,20 +283,39 @@ func asPallelBenchmark(c chan TestCase, next func() TestCase) func(*testing.B) {
 	}
```
Comment (cfarmer-fearless) [inline tests/bad-uploader/main.go:283]: Typo in function above: `asPallelBenchmark`

**Review by cfarmer-fearless — CHANGES_REQUESTED** [review]
Getting the following error when trying to run
`# command-line-arguments
.\main.go:297:8: undefined: TemplateGenerator
`

File: tests/bad-uploader/main.go (hunk 4 above)
Comment (whytheplatypus) [reply tests/bad-uploader/main.go:317]: allows the client to upload a file who's size it doesn't know until later: https://tus.io/protocols/resumable-upload#upload-defer-length

**Comment by whytheplatypus** [comment]
> Getting the following error when trying to run `# command-line-arguments .\main.go:297:8: undefined: TemplateGenerator `

Need to run it to include all files, so `go run ./...`

**Review by cfarmer-fearless — APPROVED** [review]

**Pull request merged**